File: python/sdk/languages.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Any, Dict, List, Optional
import requests


@dataclass
class Language:
    id: Optional[str] = None
    code: str = ""
    name: str = ""
    native_name: Optional[str] = None

    @staticmethod
    def from_dict(payload: Dict[str, Any]) -> "Language":
        return Language(
            id=payload.get("id"),
            code=str(payload.get("code")),
            name=str(payload.get("name")),
            native_name=payload.get("nativeName") or payload.get("native_name"),
        )

@dataclass
class LanguagesResponse:
    languages: List[Language]
    default_locale: Optional[str] = None
    raw: Optional[Any] = None  # could be list or dict

    @staticmethod
    def from_dict(payload: Any) -> "LanguagesResponse":
        # If payload is a dict with 'languages' key
        if isinstance(payload, dict):
            items = payload.get("languages") or []
            default_locale = payload.get("defaultLocale") or payload.get("default_locale")
        # If payload is a plain list
        elif isinstance(payload, list):
            items = payload
            default_locale = None
        else:
            items = []
            default_locale = None

        languages = [Language.from_dict(item) for item in items]
        return LanguagesResponse(
            languages=languages,
            default_locale=default_locale,
            raw=payload,
        )


class LanguagesClient:
    BASE_URL = "http://localhost:3005"

    def __init__(self, api_key: str, base_url: Optional[str] = None, timeout: Optional[float] = 30.0):
        if not api_key:
            raise ValueError("API key is required")
        self.api_key = api_key
        self.base_url = (base_url or self.BASE_URL).rstrip("/")
        self.timeout = timeout
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
        }

    def list_languages(self, voice_model_id: Optional[str] = None) -> LanguagesResponse:
        params = {"voice_model_id": voice_model_id} if voice_model_id else None
        resp = requests.get(
            f"{self.base_url}/languages", headers=self.headers, params=params, timeout=self.timeout
        )
        self._ensure_success(resp)
        return LanguagesResponse.from_dict(resp.json())

    def get_language_by_id(self, language_id: Optional[str] = None, id: Optional[str] = None) -> Language:
        # Support both names; prefer language_id and fall back to id
        lang_id = language_id or id
        if not lang_id:
            raise ValueError("language_id (or id) is required")
        resp = requests.get(f"{self.base_url}/languages/{lang_id}", headers=self.headers, timeout=self.timeout)
        self._ensure_success(resp)
        return Language.from_dict(resp.json())

    def _ensure_success(self, resp: requests.Response) -> None:
        if not resp.ok:
            # try to include JSON message if present
            try:
                detail = resp.json()
            except Exception:
                detail = resp.text
            raise RuntimeError(f"Request failed ({resp.status_code}): {detail}")

File: js/src/languages.ts
import fetch from "cross-fetch";

const DEFAULT_BASE_URL = "http://localhost:3005";

export type Language = {
    code: string;
    name: string;
    nativeName?: string;
};

export type LanguagesResponse = {
    languages: Language[];
    defaultLocale?: string;
};

export type ClientConfig = {
    apiKey: string;
    baseUrl?: string;
    timeoutMs?: number;
};

function buildUrl(baseUrl: string, path: string, params?: Record<string, string>): string {
    const url = new URL(path.startsWith("/") ? path : `/${path}`, baseUrl.endsWith("/") ? baseUrl : `${baseUrl}/`);
    if (params) {
        for (const [k, v] of Object.entries(params)) {
            url.searchParams.append(k, v);
        }
    }
    return url.toString();
}

export async function listLanguages({
    apiKey,
    voiceModelId,
    baseUrl = DEFAULT_BASE_URL,
    timeoutMs,
}: ClientConfig & { voiceModelId?: string }): Promise<LanguagesResponse> {
    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const params = voiceModelId ? { voice_model_id: voiceModelId } : undefined;
        const response = await fetch(
            buildUrl(baseUrl, "/languages", params),
            {
                headers: {
                    Authorization: `Bearer ${apiKey}`,
                    Accept: "application/json",
                },
                signal: controller?.signal,
            }
        );

        if (!response.ok) {
            const body = await response.text();
            throw new Error(`Languages request failed (${response.status}): ${body}`);
        }

        return response.json();
    } finally {
        if (timer) clearTimeout(timer);
    }
}

export async function getLanguageById({
    apiKey,
    id,
    baseUrl = DEFAULT_BASE_URL,
    timeoutMs,
}: ClientConfig & { id: string }): Promise<Language> {
    if (!id) throw new Error("id is required");

    const controller = typeof AbortController !== "undefined" && timeoutMs ? new AbortController() : undefined;
    const timer = controller && timeoutMs ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    try {
        const response = await fetch(buildUrl(baseUrl, `/languages/${encodeURIComponent(id)}`), {
            headers: {
                Authorization: `Bearer ${apiKey}`,
                Accept: "application/json",
            },
            signal: controller?.signal,
        });

        if (!response.ok) {
            const body = await response.text();
            throw new Error(`Languages request failed (${response.status}): ${body}`);
        }

        return response.json();
    } finally {
        if (timer) clearTimeout(timer);
    }
}
